dataflow/src/main/java/org/checkerframework/dataflow/cfg/AbstractCFGVisualizer.java

```diff
@@ -0,0 +1,455 @@
+ *
+ * @see DOTCFGVisualizer
+ * @see StringCFGVisualizer
+ */
+public abstract class AbstractCFGVisualizer<
+                A extends AbstractValue<A>, S extends Store<S>, T extends TransferFunction<A, S>>
+        implements CFGVisualizer<A, S, T> {
+
+    /**
+     * Initialized in {@link #init(Map)}. If its value is {@code true}, {@link CFGVisualizer} will
+     * return more detailed information.
+     */
+    protected boolean verbose;
+
+    /** The line separator. */
+    protected final String lineSeparator = System.lineSeparator();
+
+    /**
+     * Indicate 2 white space characters as the indentation to the elements of the {@code store}.
+     */
+    protected final String storeEntryIndent = "  ";
+
+    @Override
+    public void init(Map<String, Object> args) {
+        Object verb = args.get("verbose");
+        this.verbose =
+                verb != null
+                        && (verb instanceof String
+                                ? Boolean.getBoolean((String) verb)
+                                : (boolean) verb);
+    }
+
+    /**
+     * Generate the control flow graph.
+     *
+     * @param cfg the current control flow graph
+     * @param entry the entry block of the control flow graph
+     * @param analysis the current analysis
+     * @return the specific representation of the control flow graph, e.g., DOT, String
+     */
+    protected String generateGraph(
+            ControlFlowGraph cfg, Block entry, @Nullable Analysis<A, S, T> analysis) {
+        return visualizeGraphHeader()
+                + generateGraphHelper(cfg, entry, analysis)
+                + visualizeGraphFooter();
+    }
+
+    /**
+     * Helper method to simplify generating a control flow graph.
+     *
+     * @param cfg the control flow graph
+     * @param entry the entry block
+     * @param analysis the current analysis
+     * @return the String representation of the control flow graph
+     */
+    protected String generateGraphHelper(
+            ControlFlowGraph cfg, Block entry, @Nullable Analysis<A, S, T> analysis) {
+        Set<Block> visited = new HashSet<>();
+        StringBuilder sbDigraph = new StringBuilder();
+        Queue<Block> workList = new ArrayDeque<>();
+        Block cur = entry;
+        visited.add(entry);
+        while (cur != null) {
+            handleSuccessorsHelper(cur, visited, workList, sbDigraph);
+            cur = workList.poll();
+        }
+        sbDigraph.append(generateNodes(visited, cfg, analysis));
+        return sbDigraph.toString();
+    }
+
+    /**
+     * Helper method called by {@link #generateGraphHelper(ControlFlowGraph, Block, Analysis)}. It
+     * checks the successors of the current block, and, if possible, adds all the successors to the
+     * work list and the visited blocks list.
+     *
+     * @param cur the current block
+     * @param visited the set of blocks that have already been visited
```

Comment:
Or are in the worklist?
If so, please update the comment here and throughout, anywhere else that `@param visited` is documented.

Reply:
Thanks for the comments!
Here is what I think. We traverse the control flow graph to get all the blocks, it's like breadth first search. So `visited` is to store all the blocks that have been visited, `workList` is to store the blocks that will be processed in the future ( to check the successors ).
So I think maybe for now, it is suitable.

Reply:
My guess from reading the code is that visited is a superset of workList.
For example, addBlock always adds to visited whenever adding to workList.

You are saying that my understanding is incorrect.  Can you point out when visited does *not* contain a value that is on the worklist?
Also, if that is not the case, then workList may contain duplicates and the code is buggy. It may process nodes more than once, because the code that traverses workList does not check for duplicates.

Reply:
Sorry for the misunderstanding, I think you are right, `visited` is to log all the blocks.
Can we say "the set of blocks that are in the workList" ? Since the number of blocks in visited is always larger than the number of blocks in workList. Also, the entry block never exists in workList. 

Is it suitable to say "the set of blocks that have already logged"?

Reply:
If there is a clear, documented concept of "logged", that would be OK.
But I think it's clearer to say "the set of blocks that have already been visited or are in the worklist".
Do you have an alternate suggestion?

Reply:
That sounds great! I will do it.